File: Android/app/src/main/java/com/ditto/guides/services/DittoService.kt
package com.ditto.guides.services

import android.content.Context
import com.ditto.guides.models.AppConfig
import com.ditto.guides.models.Planet
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import live.ditto.Ditto
import live.ditto.DittoError
import live.ditto.DittoIdentity
import live.ditto.DittoLogLevel
import live.ditto.DittoLogger
import live.ditto.DittoStoreObserver
import live.ditto.DittoSyncSubscription
import live.ditto.android.DefaultAndroidDittoDependencies

interface DittoService {
    val ditto: Ditto?
    fun getPlanets(): Flow<List<Planet>>
    suspend fun updatePlanet(planet: Planet)
    suspend fun addPlanet(planet: Planet)
    suspend fun archivePlanet(planetId: String)
}

class DittoServiceImp(
    appConfig: AppConfig,
    context: Context,
    private val errorService: ErrorService
) : DittoService {

    override var ditto: Ditto? = null
    private var subscription: DittoSyncSubscription? = null
    private var planetObserver: DittoStoreObserver? = null

    init {
        try {
            DittoLogger.minimumLogLevel = DittoLogLevel.DEBUG
            val androidDependencies = DefaultAndroidDittoDependencies(context)

            //
            // CustomUrl is used because Connector is in Private Preview
            // and uses a different cluster than normal production
            //
            val customAuthUrl = "https://${appConfig.endpointUrl}"
            val webSocketUrl = "wss://${appConfig.endpointUrl}"

            //
            // TODO remove when Connector is out of Private Preview
            //
            val identity = DittoIdentity.OnlinePlayground(
                androidDependencies,
                appConfig.appId,
                appConfig.authToken,
                false,
                customAuthUrl
            )
            this.ditto = Ditto(androidDependencies, identity)
            ditto?.updateTransportConfig { config ->
                config.connect.websocketUrls.add(webSocketUrl)
            }

            setupSubscriptions()

        } catch (e: DittoError) {
            errorService.showError("Failed to initialize Ditto: ${e.message}")
        }
    }

    /**
     * Sets up the initial subscription to the planets collection in Ditto.
     * This subscription ensures that changes to the planets collection are synced
     * between the local Ditto store and the MongoDB Atlas database.
     *
     * The subscription:
     * - Queries all non-archived planets (isArchived = false)
     * - Starts the sync process with the Ditto cloud
     * - Maintains the subscription until the app is terminated or sync is stopped
     *
     * Note: This is different from the observer which notifies of data changes.
     * This subscription is responsible for the actual data synchronization.
     */
    private fun setupSubscriptions() {
        try {
            this.ditto?.let {
                this.subscription = it.sync.registerSubscription(
                    """
                    SELECT *
                    FROM planets
                    WHERE isArchived = :isArchived
                    """, mapOf("isArchived" to false)
                )
                it.startSync()
            }
        } catch (e: Exception) {
            errorService.showError("Failed to setup ditto subscription: ${e.message}")
        }
    }

    /**
     * Creates a Flow that observes and emits changes to the planets collection in Ditto.
     *
     * This method:
     * - Sets up a live query observer for the planets collection
     * - Emits a new list of planets whenever the data changes
     * - Automatically cleans up the observer when the flow is cancelled
     *
     * The query:
     * - Returns all non-archived planets (isArchived = false)
     * - Orders the results by distance from the sun
     * - Transforms raw Ditto data into Planet objects
     *
     * @return Flow<List<Planet>> A flow that emits updated lists of planets
     * whenever changes occur in the collection
     *
     */
    override fun getPlanets(): Flow<List<Planet>> = callbackFlow {
        try {
            ditto?.let {
                planetObserver = it.store.registerObserver(
                    """
                SELECT *
                FROM planets
                WHERE isArchived = :isArchived
                ORDER BY orderFromSun
            """, mapOf("isArchived" to false)
                ) { results ->
                    val planets = results.items.map { item ->
                        Planet.fromMap(item.value)
                    }
                    trySend(planets)
                }
            }

            // Clean up the observer when the flow is cancelled
            awaitClose {
                planetObserver?.close()
                planetObserver = null
            }
        } catch (e: Exception) {
            errorService.showError("Failed to setup observer for getting planets: ${e.message}")
        }
    }

    override suspend fun updatePlanet(planet: Planet) {
        try {
            ditto?.store?.execute(
                """
                UPDATE planets
                SET hasRings = :hasRings,
                    isArchived = :isArchived,
                    mainAtmosphere = :atmosphere,
                    name = :name,
                    orderFromSun = :orderFromSun,
                    surfaceTemperatureC = :temperature
                WHERE planetId = :planetId
                """,
                mapOf(
                    "hasRings" to planet.hasRings,
                    "isArchived" to planet.isArchived,
                    "atmosphere" to planet.mainAtmosphere,
                    "name" to planet.name,
                    "orderFromSun" to planet.orderFromSun,
                    "planetId" to planet.planetId,
                    "temperature" to mapOf(
                        "max" to planet.surfaceTemperatureC.max,
                        "mean" to planet.surfaceTemperatureC.mean,
                        "min" to planet.surfaceTemperatureC.min
                    )
                )
            )
        } catch (e: Exception) {
            errorService.showError("Failed to update planet: ${e.message}")
        }
    }

    override suspend fun addPlanet(planet: Planet) {
        try {
            ditto?.store?.execute(
                """
                INSERT INTO planets DOCUMENTS (:newPlanet)
                """,
                mapOf(
                    "newPlanet" to mapOf(
                        "_id" to planet.id,
                        "hasRings" to planet.hasRings,
                        "isArchived" to planet.isArchived,
                        "mainAtmosphere" to planet.mainAtmosphere,
                        "name" to planet.name,
                        "orderFromSun" to planet.orderFromSun,
                        "planetId" to planet.planetId,
                        "surfaceTemperatureC" to mapOf(
                            "max" to planet.surfaceTemperatureC.max,
                            "mean" to planet.surfaceTemperatureC.mean,
                            "min" to planet.surfaceTemperatureC.min
                        )
                    )
                )
            )
        } catch (e: Exception) {
            errorService.showError("Failed to add planet: ${e.message}")
        }
    }

    override suspend fun archivePlanet(planetId: String) {
        try {
            ditto?.store?.execute(
                """
                UPDATE planets
                    SET isArchived = :isArchived
                    WHERE planetId = :planetId
                """,
                mapOf(
                    "isArchived" to true,
                    "planetId" to planetId,
                )
            )
        } catch (e: Exception) {
            errorService.showError("Failed to add planet: ${e.message}")
        }
    }
}

File: Dotnet/Services/DittoService.cs
using Guides.Models;
using DittoSDK;

namespace Guides.Services;

public interface IDataService
{
    void GetPlanets(Action<Planet[]> callback);
    Task UpdatePlanetAsync(Planet planet);
    Task AddPlanetAsync(Planet planet);
    Task ArchivePlanetAsync(string planetId);
}

public class DittoService : IDataService
{
    private readonly AppConfig _appConfig;
    private readonly ErrorService _errorService;
    private readonly Ditto? _ditto;
    private DittoSyncSubscription? _subscription = null;
    private DittoStoreObserver? _planetObserver = null;
    
    public DittoService(AppConfig appConfig, ErrorService errorService)
    {
        _appConfig = appConfig;
        _errorService = errorService;
        
        try 
        {
            //setup Ditto logging
            DittoLogger.SetMinimumLogLevel(DittoLogLevel.Debug);
            
            //
            // CustomUrl is used because Connector is in Private Preview
            // and uses a different cluster than normal production
            //
            var customAuthUrl = $"https://{appConfig.EndpointUrl}";
            var webSocketUrl = $"wss://{appConfig.EndpointUrl}";
            
            //
            // TODO remove when Connector is out of Private Preview
            // https://docs.ditto.live/sdk/latest/install-guides/c-sharp#initializing-ditto
            
            //
            //TODO once bug is fix switch enableDittoCloud to false and set customAuthUrl
            //
            var identity = DittoIdentity 
                .OnlinePlayground( 
                    appConfig.AppId, 
                    appConfig.AuthToken, 
                    true);

            this._ditto = new Ditto(identity);
            _ditto.DisableSyncWithV3(); // remove this once using customAuthUrl
            
            // TODO add in transport configuration
            // this._ditto?.TransportConfig.Connect.WebsocketUrls.Add(webSocketUrl);
            
            SetupSubscriptions();
        }
        catch (Exception ex)
        {
            _errorService.ShowError($"Failed to initialize Ditto: {ex.Message}");
        }
    }

    /// <summary>
    /// Sets up the initial subscription to the planets collection in Ditto.
    /// </summary>
    /// <remarks>
    /// This subscription ensures that changes to the planets collection are synced 
    /// between the local Ditto store and the MongoDB Atlas database.
    /// 
    /// The subscription:
    /// <list type="bullet">
    /// <item><description>Queries all non-archived planets (isArchived = false)</description></item>
    /// <item><description>Starts the sync process with the Ditto cloud</description></item>
    /// <item><description>Maintains the subscription until the app is terminated or sync is stopped</description></item>
    /// </list>
    /// 
    /// </remarks>
    /// <seealso href="https://docs.ditto.live/sdk/latest/sync/syncing-data#creating-subscriptions">Ditto Sync Documentation</seealso>
    private void SetupSubscriptions()
    {
        try
        {
            if (_ditto == null || _subscription != null) return;
            //create a subscription to sync the planets collection 
            _subscription = _ditto.Sync.RegisterSubscription(
                "SELECT * FROM planets WHERE isArchived = :isArchived", 
                new Dictionary<string, object> { { "isArchived", false } }
            );
            _ditto.StartSync();
        }
        catch (Exception ex)
        {
            _errorService.ShowError($"Failed to Setup Subscription: {ex.Message}");
        }
    }

    /// <summary>
    /// Registers an observer for the planets collection and provides real-time updates through a callback.
    /// </summary>
    /// <remarks>
    /// This method:
    /// <list type="bullet">
    /// <item><description>Sets up a live query observer for the planets collection</description></item>
    /// <item><description>Returns only non-archived planets (isArchived = false)</description></item>
    /// <item><description>Orders results by distance from the sun</description></item>
    /// <item><description>Automatically updates when changes occur</description></item>
    /// </list>
    /// </remarks>
    /// <param name="callback">Action that receives updates when the planets collection changes</param>
    /// <seealso href="https://docs.ditto.live/sdk/latest/crud/observing-data-changes#store-observer-with-query-arguments">Ditto Observer Documentation</seealso>
    public void GetPlanets(Action<Planet[]> callback)
    {
        try {
            if (_ditto != null && _subscription != null)
            {
                var queryArguments = new Dictionary<string, object>(){{"isArchived", false }};

                // Register an observer to listen for changes to any documents in the planets collection
                _planetObserver = _ditto.Store.RegisterObserver(
                    "SELECT * FROM planets WHERE isArchived = :isArchived ORDER BY orderFromSun", 
                    queryArguments,
                    result =>
                    {
                        var planets = new List<Planet>();
                        result.Items.ForEach(item =>
                        {
                            planets.Add(Planet.FromDictionary(item.Value));
                        });    
                       callback(planets.ToArray()); 
                    } );
            }
        } catch (Exception ex) {
            _errorService.ShowError($"Error with GetPlanetAsync: {ex.Message}");
        }
    }

    /// <summary>
    /// Updates an existing planet's properties in the Ditto store.
    /// </summary>
    /// <remarks>
    /// This method:
    /// <list type="bullet">
    /// <item><description>Updates all mutable fields of the planet</description></item>
    /// <item><description>Maintains the planet's ID and references</description></item>
    /// <item><description>Triggers a sync with other devices</description></item>
    /// </list>
    /// </remarks>
    /// <param name="planet">The Planet object containing the updated values</param>
    /// <returns>A task representing the asynchronous operation</returns>
    /// <exception cref="Exception">Thrown when the update operation fails</exception>
    /// <seealso href="https://docs.ditto.live/sdk/latest/crud/update#updating">Ditto Update Documentation</seealso>
    public async Task UpdatePlanetAsync(Planet planet)
    {
        try
        {
            if (_ditto != null)
            {
                var updateArguments = planet.ToDictionary(); 
                var results = await _ditto.Store.ExecuteAsync(
                    "UPDATE planets SET hasRings = :hasRings, isArchived = :isArchived, mainAtmosphere = :atmosphere, name = :name, orderFromSun = :orderFromSun, surfaceTemperatureC = :temperature  WHERE planetId = :planetId", 
                    updateArguments);
                if (results.Items.Count == 0)
                {
                    _errorService.ShowError("Error: Planet Update didn't return any results.");
                }
            }
        }
        catch (Exception ex)
        {
            _errorService.ShowError($"Failed to update planet: {ex.Message}");
            throw;
        }
    }

    /// <summary>
    /// Creates a new planet document in the Ditto store.
    /// </summary>
    /// <remarks>
    /// This method:
    /// <list type="bullet">
    /// <item><description>Creates a new document in the planets collection</description></item>
    /// <item><description>Assigns the provided ID and properties</description></item>
    /// <item><description>Sets initial isArchived status to false</description></item>
    /// <item><description>Triggers a sync with other devices</description></item>
    /// </list>
    /// </remarks>
    /// <param name="planet">The new Planet object to be added</param>
    /// <returns>A task representing the asynchronous operation</returns>
    /// <exception cref="Exception">Thrown when the insert operation fails</exception>
    /// <seealso href="https://docs.ditto.live/sdk/latest/crud/create#creating-documents">Ditto Create Documentation</seealso>
    public async Task AddPlanetAsync(Planet planet)
    {
        try
        {
            if (_ditto != null)
            {
                var arguments = planet.ToDictionary(); 
                var insertArguments = new Dictionary<string, object>{{"newPlanet", arguments}};
                var results = await _ditto.Store.ExecuteAsync(
                    "INSERT INTO planets DOCUMENTS (:newPlanet)", 
                    insertArguments);
                if (results.Items.Count == 0)
                {
                    _errorService.ShowError("Error: Adding Planet didn't return any results.");
                }
            }
        }
        catch (Exception ex)
        {
            _errorService.ShowError($"Failed to update planet: {ex.Message}");
            throw;
        }
    }

    /// <summary>
    /// Archives a planet by setting its isArchived flag to true.
    /// </summary>
    /// <remarks>
    /// This method implements the 'Soft-Delete' pattern:
    /// <list type="bullet">
    /// <item><description>Marks the planet as archived instead of deleting it</description></item>
    /// <item><description>Removes it from active queries and views</description></item>
    /// <item><description>Maintains the data for historical purposes</description></item>
    /// <item><description>Triggers a sync with other devices</description></item>
    /// </list>
    /// </remarks>
    /// <param name="planetId">The unique identifier of the planet to archive</param>
    /// <returns>A task representing the asynchronous operation</returns>
    /// <exception cref="Exception">Thrown when the archive operation fails</exception>
    /// <seealso href="https://docs.ditto.live/sdk/latest/crud/delete#soft-delete-pattern">Ditto Soft Delete Documentation</seealso>
    public async Task ArchivePlanetAsync(string planetId)
    {
        try
        {
            if (_ditto != null)
            {
                var updateArguments = new Dictionary<string, object>
                {
                    { "isArchived", true },
                    { "planetId", planetId }
                };
                var results = await _ditto.Store.ExecuteAsync(
                    " UPDATE planets SET isArchived = :isArchived WHERE planetId = :planetId", 
                    updateArguments);
                if (results.Items.Count == 0)
                {
                    _errorService.ShowError("Error: Planet Update didn't return any results.");
                }
            }
        }
        catch (Exception ex)
        {
            _errorService.ShowError($"Failed to update planet: {ex.Message}");
            throw;
        }
    }
}